src/App.tsx

```diff
@@ -15,47 +15,140 @@ export const goodsFromServer = [
+      visibleGoods = visibleGoods.sort((a, b) => a.localeCompare(b));
+      break;
+
+    case SortType.LENGTH:
+      visibleGoods = visibleGoods.sort((a, b) => a.length - b.length);
+      break;
+
+    default:
+      //
+  }
+
+  if (isReversed) {
+    return visibleGoods.reverse();
+  }
+
+  return visibleGoods;
+}
+
 export const App: React.FC = () => {
+  const [appState, setAppState] = React.useState({
+    sortType: SortType.NONE,
+    isReversed: false,
+  });
+
+  const handelSortAlphbet = () => {
+    setAppState(prevState => {
+      return {
+        ...prevState,
+        sortType: SortType.ALPABET,
+      };
+    });
+  };
+
+  const handelSortLength = () => {
+    setAppState(prevState => {
+      return {
+        ...prevState,
+        sortType: SortType.LENGTH,
+      };
+    });
+  };
+
+  const handelReverse = () => {
+    setAppState((prevState) => {
+      return {
+        ...prevState,
+        isReversed: !prevState.isReversed,
+      };
+    });
+  };
+
+  const handelReset = () => {
+    setAppState({
+      isReversed: false,
+      sortType: SortType.NONE,
+    });
+  };
+
   return (
     <div className="section content">
       <div className="buttons">
         <button
           type="button"
-          className="button is-info is-light"
+          className={appState.sortType === SortType.ALPABET
+            ? 'button is-info'
+            : 'button is-info is-light'}
+          onClick={handelSortAlphbet}
         >
           Sort alphabetically
         </button>
 
         <button
           type="button"
-          className="button is-success is-light"
+          className={appState.sortType === SortType.LENGTH
+            ? 'button is-success'
+            : 'button is-success is-light'}
+          onClick={handelSortLength}
         >
           Sort by length
         </button>
 
         <button
           type="button"
-          className="button is-warning is-light"
+          className={appState.isReversed
+            ? 'button is-warning'
+            : 'button is-warning is-light'}
+          onClick={handelReverse}
         >
           Reverse
         </button>
 
-        <button
-          type="button"
-          className="button is-danger is-light"
-        >
-          Reset
-        </button>
+        {(appState.isReversed || appState.sortType !== SortType.NONE) && (
+          <button
+            type="button"
+            className="button is-danger is-light"
+            onClick={handelReset}
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
         <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
+          {getReorderedGoods(
+            goodsFromServer,
+            appState,
+          ).map(good => (
```

Comment:
create variable. you can also memoize it (useMemo) so you will have a performance increase, wherein the current setup you always run the function getReorderedGoods, even if the order did not change

Reply:
Variable created for storing sorted goods and React useMemo was used for optimized performance. Also the App states was NOT modified if the prevState was unchange. So that clicking any sort button more than once will not re-render the App for optimized performacne.